File: scripts/self_heal.py
#!/usr/bin/env python3
"""Targeted, ordered repairs for html2md.

Each step is idempotent and re-runs the healthcheck after changes.
Exit 0 if the healthcheck eventually passes, regardless of whether any files changed.
"""
from __future__ import annotations

import subprocess
import sys


def sh(cmd: list[str], label: str) -> bool:
    """Run a command, return True on success."""
    print(f"\n$ {' '.join(cmd)}  # {label}")
    result = subprocess.run(cmd)
    return result.returncode == 0


def has_changes() -> bool:
    """Return True if the working tree has uncommitted changes."""
    result = subprocess.run(
        ["git", "status", "--porcelain"],
        capture_output=True,
        text=True,
    )
    return bool(result.stdout.strip())


def passes_healthcheck() -> bool:
    """Return True if scripts/healthcheck.py exits 0."""
    # We call the healthcheck script using the same python interpreter
    result = subprocess.run([sys.executable, "scripts/healthcheck.py"])
    return result.returncode == 0


def fix_dependencies() -> bool:
    """Install/upgrade dependencies and dev tools."""
    ok = True
    ok &= sh([sys.executable, "-m", "pip", "install", "-e", "."], "editable install")
    ok &= sh([sys.executable, "-m", "pip", "install", "pytest", "ruff"], "install dev tools")
    return ok


def fix_lint() -> bool:
    """Run ruff linting and formatting fixes."""
    ok = True
    # We target src/ and tests/ explicitly to match typical usage,
    # but "." is also fine if ruff defaults are set.
    # Using "." to catch everything including scripts/.
    ok &= sh([sys.executable, "-m", "ruff", "check", "--fix", "."], "ruff fix")
    ok &= sh([sys.executable, "-m", "ruff", "format", "."], "ruff format")
    return ok


def fix_install(force: bool = False) -> bool:
    """Force reinstall if needed."""
    flags = ["--upgrade"]
    if force:
        flags.append("--force-reinstall")
    return sh(
        [sys.executable, "-m", "pip", "install"] + flags + ["-e", "."],
        "reinstall deps" if force else "upgrade deps",
    )


def main() -> int:
    """Run repair steps sequentially, but stop if healthcheck starts passing."""
    if passes_healthcheck():
        print("\n=== Healthcheck already passes. ===")
        return 0

    print("\n=== Step 1: Install Dependencies ===")
    fix_dependencies()
    if passes_healthcheck():
        print("\n=== Fixed by installing dependencies. ===")
        # If fixed, we exit 0 regardless of changes
        return 0

    print("\n=== Step 2: Lint/Format Auto-fix ===")
    fix_lint()
    if passes_healthcheck():
        print("\n=== Fixed by linting/formatting. ===")
        return 0

    print("\n=== Step 3: Upgrade Dependencies ===")
    fix_install(force=False)
    if passes_healthcheck():
        print("\n=== Fixed by upgrading dependencies. ===")
        return 0

    print("\n=== Step 4: Force Reinstall ===")
    fix_install(force=True)
    if passes_healthcheck():
        print("\n=== Fixed by force reinstall. ===")
        return 0

    print("\n=== Self-heal could not fix all issues. ===", file=sys.stderr)
    return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/self-heal.mjs
#!/usr/bin/env node
/* Targeted, ordered repairs. Each step is idempotent and re-runs healthcheck.
   Exit 0 only if a repair produced a passing healthcheck and a non-empty diff.
*/
import { execSync, spawnSync } from "node:child_process";
import { existsSync, readFileSync } from "node:fs";

const sh = (cmd, opts = {}) => {
  console.log(`\n$ ${cmd}`);
  return execSync(cmd, { stdio: "inherit", ...opts });
};
const trySh = (cmd, opts = {}) => {
  try { sh(cmd, opts); return true; } catch { return false; }
};
const changed = () => {
  const out = spawnSync("git", ["status", "--porcelain"], { encoding: "utf8" });
  return (out.stdout || "").trim().length > 0;
};
const passHealth = () => {
  try { sh("node scripts/healthcheck.mjs"); return true; } catch { return false; }
};
const readPackageJson = () => {
  if (!existsSync("package.json")) return {};
  const contents = readFileSync("package.json", "utf8").trim();
  if (!contents) return {};
  try {
    return JSON.parse(contents);
  } catch (error) {
    console.warn(`Ignoring unreadable package.json: ${error.message}`);
    return {};
  }
};
const packageJson = readPackageJson();
const scripts = packageJson.scripts && typeof packageJson.scripts === "object" ? packageJson.scripts : {};
const hasScript = (name) => typeof scripts[name] === "string";
const hasTypeScriptConfig = () => existsSync("tsconfig.json") || existsSync("jsconfig.json");

let fixed = false;

// 1) Lint/format. This repository is not a pnpm workspace, so use root package
// scripts directly when they exist rather than workspace-only `pnpm -w` flags.
if (hasScript("lint")) trySh("pnpm run lint -- --fix");
if (hasScript("format")) trySh("pnpm run format");
if (passHealth()) fixed = fixed || changed();

// 2) Snapshot updates (only when a Vitest-backed test script is configured).
if (!passHealth() && hasScript("test") && /vitest/.test(scripts.test)) {
  trySh("pnpm exec vitest -u");
  if (passHealth()) fixed = fixed || changed();
}

// 3) Type acquisition (only for actual TypeScript projects).
if (!passHealth() && hasScript("typecheck") && hasTypeScriptConfig()) {
  trySh("pnpm dlx typesync --save-dev");
  // In case typesync suggests @types/node et al.
  trySh("pnpm install");
  if (passHealth()) fixed = fixed || changed();
}

// 4) Lockfile repair (only when this repository already tracks a pnpm lockfile).
if (!passHealth() && existsSync("pnpm-lock.yaml")) {
  // Try a clean install + re-resolve.
  trySh("pnpm install");
  if (!passHealth()) {
    // Last resort: conservatively refresh dependency versions allowed by package.json.
    trySh("pnpm up --interactive=false");
  }
  if (passHealth()) fixed = fixed || changed();
}

// 5) Known generators (icons/docs), if present.
if (!passHealth()) {
  if (existsSync("scripts/update-icon-docs.mjs")) {
    trySh("node scripts/update-icon-docs.mjs");
  }
  if (existsSync("scripts/verify-static.mjs")) {
    trySh("node scripts/verify-static.mjs");
  }
  if (passHealth()) fixed = fixed || changed();
}

if (fixed) {
  console.log("Self-heal successful and produced a diff.");
  process.exit(0);
} else {
  console.log("Self-heal failed or no changes were needed.");
  process.exit(1);
}
